core/src/main/java/com/graphhopper/routing/util/HeadingEdgeFilter.java

```diff
@@ -18,6 +18,13 @@ public HeadingEdgeFilter(EdgeFilter directedEdgeFilter, double heading, GHPoint
 
     @Override
     public boolean accept(EdgeIteratorState edgeState) {
-        return HeadingResolver.isHeadingNearlyParallel(edgeState, directedEdgeFilter, heading, pointNearHeading);
+        final double tolerance = 30;
+        double headingOfEdge = HeadingResolver.getHeadingOfGeometryNearPoint(edgeState, pointNearHeading);
+        if (Double.isNaN(headingOfEdge))
+            // this edge is too far away. we do not accept it.
+            return false;
+        // the edge is not directed. we accept the edge if either of the two directions roughly has the right heading
+        return directedEdgeFilter.accept(edgeState) && Math.abs(headingOfEdge - heading) < tolerance ||
+                directedEdgeFilter.accept(edgeState.detach(true)) && Math.abs((headingOfEdge + 180) % 360 - heading) < tolerance;
```

Comment:
As accept is likely more expensive I would swap this with the tolerance check and so we can avoid it in cases the tolerance does not fit?

Reply:
Sure. Done.